### PR TITLE
Support Deis Workflow (V2) client

### PR DESCRIPTION
**Pull request by mattk42**
This change adds support to install the Deis Workflow V2 client. The usage of the client has remained the same as in V1, but it is now installed using a different script.

Currently V2 is in beta and the client install script only supports installing the latests release of the V2 client. So for now this will install the latest V2 client if the "cli_version" option is set to 2. This will likely change once the installation script is out of beta and supports installing other clients.

I made no change to the tests, as they do not utilize the actual client and simply return a true for the shell out that is called.

This change has been tested against a working Deis V2 environment and the deploy succeeds as expected.

### CODE REVIEW

**Comment by Joshua-Anderson** [comment]
This is a pretty brittle solution, because only string `2` matches. Deis `2.1.0` comes out next week, and that won't work. Right now `http://deis.io/deis-cli/install-v2.sh` doesn't support different versions. However, you can directly download the CLI with a url in this format: `https://storage.googleapis.com/workflow-cli/v2.0.0/deis-v2.0.0-linux-amd64`

**Comment by mattk42** [comment]
Yeah, I realized that when I wrote it but Deis was still quite a ways away from a release and the installer wasn't finalized yet. But now that Deis is getting to the 2.1.0 release, sounds like the installer script should be updated soon and then I can update this PR to pass in the version argument.

I think it wold be best to hold off on merging this until that is finalized. I will comment here again once that has been done.

**Comment by Joshua-Anderson** [comment]
Deis v2.1 has been released and the script updated to take the version from the first arguement.

**Comment by mattk42** [comment]
Thanks @Joshua-Anderson. I went ahead and updated the plugin to parse the version and determine which type of install to do. Still defaults to a v1 client since I don't want to break anyone currently using it.

**Comment by Joshua-Anderson** [comment]
LGTM. Thanks for updating this!

**Comment by Joshua-Anderson** [comment]
Ping @BanzaiMan When you have a chance, could you possibly review this? It would be great to have dpl support deis v2. Thanks!

**Comment by BanzaiMan** [comment]
Sorry for the delay. Could you run a deployment with this code and link to it?

Thank you.

**Comment by mattk42** [comment]
@BanzaiMan Sorry for my delay as well. Here is a deploy I ran to a V2 cluster. 

https://travis-ci.org/mattk42/dpl-test/builds/150772637

**Comment by BanzaiMan** [comment]
@mattk42 Cool. Could you bring your branch up-to-date with the current `master`? Thanks.

**Comment by mattk42** [comment]
I have been using this branch for a while without merging in master, so I hadn't noticed this before, but it seems that #479 is actually broken. The PIPESTATUS array is a special bash thing, and this is using a standard shell. This causes all builds to exit as if they failed with `sh: 1: Bad substitution`. 

@BanzaiMan Since this is already in master, would you prefer a separate PR to fix that? For reference, this is a build after the merge that is seeing this: https://travis-ci.org/mattk42/dpl-test/builds/150953074

**Comment by mattk42** [comment]
@BanzaiMan I went ahead and opened #491 for that issue. I also ran a deployment with these two PRs merged together. Results are here: https://travis-ci.org/mattk42/dpl-test/builds/151029281

**Comment by jschneiderhan** [comment]
Hi there! Is there anything I can do to help out with this issue? It would be great to get Deis v2 support merged.

**Comment by mattk42** [comment]
@BanzaiMan Anything you need from us to get this merged in?

**Comment by jschneiderhan** [comment]
My latest attempt at deploying resulted in a `Error: You do not have permission to perform this action`:

```
Building dpl gem locally with source mattk42/dpl and branch deis-v2

$ git clone https://github.com/mattk42/dpl mattk42/dpl
Cloning into 'mattk42/dpl'...
remote: Counting objects: 6822, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 6822 (delta 2), reused 0 (delta 0), pack-reused 6819
Receiving objects: 100% (6822/6822), 4.90 MiB | 0 bytes/s, done.
Resolving deltas: 100% (4106/4106), done.
Checking connectivity... done.


$ git checkout deis-v2
Branch deis-v2 set up to track remote branch deis-v2 from origin.
Switched to a new branch 'deis-v2'

$ rvm 1.9.3 --fuzzy do ruby -S gem build dpl.gemspec
WARNING:  pessimistic dependency on rspec (~> 3.0.0, development) may be overly strict
  if rspec is semantically versioned, use:
    add_development_dependency 'rspec', '~> 3.0', '>= 3.0.0'
WARNING:  open-ended dependency on rspec-its (>= 0, development) is not recommended
  if rspec-its is semantically versioned, use:
    add_development_dependency 'rspec-its', '~> 0'
WARNING:  open-ended dependency on rake (>= 0, development) is not recommended
  if rake is semantically versioned, use:
    add_development_dependency 'rake', '~> 0'
WARNING:  open-ended dependency on coveralls (>= 0, development) is not recommended
  if coveralls is semantically versioned, use:
    add_development_dependency 'coveralls', '~> 0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: dpl
  Version: 1.8.23.travis.623.1
  File: dpl-1.8.23.travis.623.1.gem






Successfully installed dpl-1.8.23.travis.623.1
1 gem installed

Installing deploy dependencies
Fetching: git-1.3.0.gem (100%)
Successfully installed git-1.3.0
1 gem installed
+VERSION=v2.7.0
+set -eo pipefail -o nounset
+PROGRAM=deis
++tr '[:upper:]' '[:lower:]'
++uname
+PLATFORM=linux
++uname -m
+ARCH=x86_64
+DEIS_BIN_URL_BASE=https://storage.googleapis.com/workflow-cli-release
+'[' x86_64 == x86_64 ']'
+ARCH=amd64
+check_platform_arch
+local 'supported=linux-amd64 darwin-amd64'
+grep -q linux-amd64
+tr ' ' '\n'
+echo 'linux-amd64 darwin-amd64'
+DEIS_CLI=deis-v2.7.0-linux-amd64
+DEIS_CLI_PATH=deis-v2.7.0-linux-amd64
+'[' v2.7.0 '!=' stable ']'
+DEIS_CLI_PATH=v2.7.0/deis-v2.7.0-linux-amd64
+echo 'Downloading deis-v2.7.0-linux-amd64 From Google Cloud Storage...'
Downloading deis-v2.7.0-linux-amd64 From Google Cloud Storage...
+curl -Ls -o deis https://storage.googleapis.com/workflow-cli-release/v2.7.0/deis-v2.7.0-linux-amd64
+chmod +x deis
+cat

deis is now available in your current directory.

To learn more about deis, execute:

    $ ./deis --help


Preparing deploy
Logged in as travis
Configuration file written to /home/travis/.deis/client.json
Error: You do not have permission to perform this action.

Removing testing-worker-linux-docker-333bfc03-3423-linux-5 SSH Key...
Error: Not Found
No stash found.
Application could not be verified.
failed to deploy
```

**Comment by jschneiderhan** [comment]
Once I had my permissions setup correctly, I was able to deploy, but after the deployment, the Deis builder rejected the push and caused the build to fail (details: https://gist.github.com/jschneiderhan/8b39d6e7fe3a427ddbbeac56fcdab4dd).

If I change the git depth to 99999 in my .travis.yml file, it works! It feels kind of hacky to do that, though. Does anyone have any suggestions on what to do in this case?

**Comment by bacongobbler** [comment]
FYI the install-v2.sh script [should now handle version arguments](https://github.com/deis/deis.io/blob/5ff57593ffc0f4fc5cac6a7ffa0fecaebdc05482/deis-cli/install-v2.sh#L12-L13) :)

File: lib/dpl/provider/deis.rb [hunk 1]
```diff
@@ -87,7 +102,7 @@ def push_app
       end
 
       def run(command)
-        unless context.shell "deis run -- #{command}"
+        unless context.shell "./deis run -- #{command}"
```
Comment (BanzaiMan) [inline lib/dpl/provider/deis.rb]: A corresponding change is required for the spec.

**Pull request merged**